### PR TITLE
Resolve paths based on location of executable

### PR DESCRIPTION
**Pull request by partouf**
These changes remove the runtime reliance on where the build was done (or installed to when using OBJ_BPREFIX= in the make command).
Also adds the possibility to change tmp folder by setting TMPDIR in the environment variables before calling the executable.

This might of course break some environments (where custom paths are used far away from the executables), so I'm open to suggestions on how to still support that if relevant.

I'm not a C developer, so I'm open to changing things up to what's standard.

I tried to make sure there were no memleaks from the code that I added, but then it got kind of messy, so I refactored the code into separate functions.

Related Compiler-Explorer issue https://github.com/compiler-explorer/misc-builder/issues/3

### CODE REVIEW

**Comment by katef** [comment]
Thank you for the work here, this is lovely.

The question I'm asking myself is fundamentally, do we want tcc to depend on finding its own location? It is something I had gone to the trouble of trying to avoid, when I worked on the env variable stuff here. But I don't see a good reason for why not. So sure, let's try living with it, and see how it goes.

File: tcc/src/main.c [hunk 1]
```diff
@@ -67,38 +67,44 @@ main_end(void)
...
+		{ "PREFIX_INCLUDE", path_join(root_path, "include")        },
+		{ "PREFIX_MAN",     path_join(root_path, "man")            },
+		{ "PREFIX_TSPEC",   path_join(root_path, "share/tspec")    },
+		{ "PREFIX_STARTUP", path_join(root_path, "lib/tcc/startup")},
+		{ "PREFIX_ENV",     path_join(root_path, "lib/tcc/env")    },
+		{ "PREFIX_API",     path_join(root_path, "lib/tcc/api")    },
+		{ "PREFIX_LPI",     path_join(root_path, "lib/tcc/lpi")    },
+		{ "PREFIX_SYS",     path_join(root_path, "lib/tcc/sys")    },
+		{ "PREFIX_MAP",     path_join(root_path, "lib/tcc/map")    },
+	};
```
Comment (katef) [inline tcc/src/main.c:92]: If we're hardcoding the strings for relative paths here, that means we no longer use the `PREFIX_` defines. So I think all the `-DPREFIX_*=...` arguments can go from src/Makefile. And that means their corresponding Makefile variables can go, too. And then they can go from the top-level Makefile, too.

File: tcc/src/main.c [hunk 2]
```diff
@@ -107,33 +113,72 @@ main_start(char *prog)
 		{ "MD_OSVER",       OSVER          }
 	};
 
+	for (size_t i = 0; i < sizeof b / sizeof *b; i++) {
```
Comment (katef) [inline tcc/src/main.c:116]: We don't have C99 here yet; please keep the `size_t` variable at block scope. We'll need `make bootstrap-rebuild` to pass (it should fail on this I think)

Comment (partouf) [reply tcc/src/main.c:116]: bootstrap-rebuild doesn't work for me in general, but was able to test this via CFLAGS=-std=c90, also brings up a couple of other things I probably need to fix

File: tcc/src/main.c (hunk 1 above)
Comment (partouf) [reply tcc/src/main.c:97]: Ok, will look at this tomorrow

File: tcc/src/main.c [hunk 3]
```diff
@@ -107,33 +119,76 @@ main_start(char *prog)
...
+ */
+
+static void
+main_start(char *prog, const char *rel_executable_path)
+{
 	atexit(main_end);
 
 	buffer = xmalloc(buffer_size);
-	set_progname(find_basename(prog), VERSION);
+	realpath(rel_executable_path, buffer);
```
[the remark this answers is not in the document]
Comment (partouf) [reply tcc/src/main.c:174]: Finding out this is actually a giant minefield and platform specific. For now I'll leave this the way it is, on my system path_max is 4096, and buffer_size is 5000, so maybe it's fine.

File: tcc/src/main.c (hunk 1 above)
Comment (partouf) [reply tcc/src/main.c:97]: I removed the -DPREFIX_ defines and some make arguments from the top-level Makefile. But I don't think I can remove much from the .mk files as they are used as target folder to install into. (I tried removing some but a lot of them are used)

Comment (partouf) [reply tcc/src/main.c:97]: currently rewriting some things so that tendra can compile the code as well

**Comment by katef** [comment]
Am happy to merge this as son as I can - I just want to test on a system where bootstrap-rebuild runs first. And apparently I've broken all the networking for my local VMs somehow. Sigh :)

**Pull request merged**